### ISSUE TITLE
FMA arithmetic tests failing for vectorization on daint

### ISSUE THREAD

**Issue by bcumming** [eth-cscs/arbor#568]
Some SIMD unit tests are failing when Arbor is compiled with vectorization support on Daint

To build:

```
module load daint-mc
module swap PrgEnv-cray PrgEnv-gnu
module swap gcc/7.3.0
module load CMake/3.11.4

export CRAYPE_LINK_TYPE=dynamic
export CC=`which gcc`; export CXX=`which g++`;


cmake .. -DARB_VECTORIZE=ON -DARB_ARCH=broadwell
```

The failing tests are `S/simd_value/*.arithmetic`, specifically those that compare `std::fma` with our vectorized `fma` operation.

**Comment by halfflat** [comment]
Weirdly, problem goes away when we reduce the number of template test cases. Investigation continues.

**Comment by bcumming** [comment]
Is this the same issue that we investigated on Power8?
Then the issue went away as we attempted to simplify the test, and different "optimizations" were turned on/off?

What happens if you take a set of offending values, and run them in isolation in their own test?

**Comment by halfflat** [comment]
It's a gcc compiler bug in the version 6 and 7 series: with -O3 and targetting haswell or broadwell architecture, FMA operations sometimes are given the wrong argument for the addition.

Submitted a bug report to GCC, so we'll see how that goes. Minimal test case (C99):
```
extern double fma(double, double, double) __attribute__((nothrow, leaf));
extern void abort() __attribute__((noreturn, nothrow, leaf));

__attribute((noinline)) void assert_eq0(const double* a, double b) {
   if (a[0]!=b) abort();
}

__attribute((noinline)) void run_fma(const double* u, const double* v, const double* w, double a) {
    double r[4];
    for (unsigned j = 0; j<1u; ++j) {
        for (unsigned i = 0; i<4; ++i) r[i] = fma(u[i], v[i], w[i]);
        assert_eq0(r, 50);
    }
}

int main() {
    double u[4] = {6, 6, 6, 6};
    double v[4] = {7, 7, 7, 7};
    double w[4] = {8, 8, 8, 8};

    run_fma(u, v, w, 50);
}
```

As this means we can't really trust the automatic vectorization on these architectures, I think we'll have to add `-fno-tree-vectorize` if we detect gcc version less than 8 on x86 platforms. First though we should verify that this workaround won't greatly impact performance. Note that the explicitly vectorized kernels will not be affected by this bug.

**Comment by halfflat** [comment]
GCC bug report: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=87046
Note that apparently 8.1 is also affected, but not 8.2.

**Comment by halfflat** [comment]
Tried a build with `-fno-tree-vectorize` and `-march=haswell` on a Xeon E5-2699 v3 (greina4): any differences in run time for the miniapp were less than measurement noise:

trial | -O3 | -O3 -fno-tree-vectorize
------|-----|--------------------------
1 | 4.045 | 3.970
2 | 3.987 | 3.982
3 | 3.981 | 3.980
4 | 3.982 | 3.980
5 | 3.980 | 3.979

I'll submit a patch that adds the `-fno-tree-vectorize` flag when architecture is specified on x86.

**Comment by halfflat** [comment]
Spoke too soon: `-fno-tree-vectorize` badly affects performance of explicitly vectorized code. It is not clear why. A slightly more complicated work around is required. With `ARB_VECTORIZE=ON`:

trial | -O3 | -O3 -fno-tree-vectorize
------|-----|--------------------------
1 | 3.562 | 4.133
2 | 3.551 | 4.135
3 | 3.550 | 4.137
4 | 3.551 | 4.128
5 | 3.556 | 4.122

**Comment by halfflat** [comment]
Using a `compat::fma` wrapper restores performance — will put this approach up for PR.

**Comment by halfflat** [comment]
Update on affected versions of gcc: underlying bug apparently fixed in 6.4.1, 7.3.1, 8.1.1 and 9.0:
https://gcc.gnu.org/bugzilla/show_bug.cgi?id=85597

**Comment by halfflat** [comment]
TODO: amend gcc version check in `compat::fma` to test for versions as above.
